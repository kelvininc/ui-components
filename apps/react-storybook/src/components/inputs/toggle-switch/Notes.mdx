import Readme from "@kelvininc/ui-components/docs/toggle-button-group.md?raw";

import { Markdown, Meta, ArgTypes } from "@storybook/blocks";
import * as ToggleSwitchStories from "./ToggleSwitch.stories";

<Meta of={ToggleSwitchStories} />
<Markdown>{Readme}</Markdown>
