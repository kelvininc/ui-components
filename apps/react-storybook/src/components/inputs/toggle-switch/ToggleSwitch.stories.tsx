import type { Meta, StoryFn, StoryObj } from '@storybook/react';
import { useArgs } from '@storybook/preview-api';
import { EComponentSize, KvToggleSwitch } from '@kelvininc/react-ui-components';
import { ComponentProps } from 'react';

const ToggleSwitchTemplate: StoryFn<ComponentProps<typeof KvToggleSwitch>> = args => {
	const [, updateArgs] = useArgs();
	const onCheckedChange = ({ detail: id }: CustomEvent<string>) =>
		updateArgs({
			...args,
			selectedOption: id
		});

	return <KvToggleSwitch {...args} onCheckedChange={onCheckedChange} />;
};

const meta = {
	title: 'Inputs/Toggle Switch',
	component: KvToggleSwitch,
	render: ToggleSwitchTemplate,
	argTypes: {
		onCheckedChange: {
			action: 'checkedChange'
		}
	}
} satisfies Meta<typeof KvToggleSwitch>;

export default meta;
type Story = StoryObj<typeof meta>;

export const TextButtonsState: Story = {
	args: {
		options: [
			{
				label: 'Option 1',
				key: 'opt1'
			},
			{
				label: 'Option 2',
				key: 'opt2'
			},
			{
				label: 'Option 3',
				key: 'opt3',
				disabled: true
			},
			{
				label: 'Option 4',
				key: 'opt4'
			},
			{
				label: 'Option 5',
				key: 'opt5'
			}
		],
		selectedOption: 'opt2',
		disabledButtons: {
			opt1: false,
			opt2: false,
			opt3: false,
			opt4: false,
			opt5: false
		},
		size: EComponentSize.Small
	}
};
