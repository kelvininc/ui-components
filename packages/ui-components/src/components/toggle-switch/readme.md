# _<kv-toggle-button-group>_

<!-- Auto Generated Below -->


## Properties

| Property          | Attribute          | Description                                                                 | Type                                           | Default                |
| ----------------- | ------------------ | --------------------------------------------------------------------------- | ---------------------------------------------- | ---------------------- |
| `disabled`        | `disabled`         | (optional) If `true` all toggle buttons will be disabled                    | `boolean`                                      | `false`                |
| `disabledButtons` | `disabled-buttons` | (optional) A record with the button's key and its individual disabled state | `{ [x: string]: boolean; }`                    | `{}`                   |
| `options`         | `options`          | (optional) List of toggle switch options                                    | `IToggleSwitchOption[]`                        | `[]`                   |
| `selectedOption`  | `selected-option`  | (optional) Toggle switch selected option key                                | `string`                                       | `''`                   |
| `size`            | `size`             | (optional) Sets the size for all toggle buttons                             | `EComponentSize.Large \| EComponentSize.Small` | `EComponentSize.Small` |


## Events

| Event           | Description                                                        | Type                  |
| --------------- | ------------------------------------------------------------------ | --------------------- |
| `checkedChange` | When the toggle switch selection changes, emit the requested value | `CustomEvent<string>` |


## Shadow Parts

| Part                               | Description                  |
| ---------------------------------- | ---------------------------- |
| `"toggle-switch-option-container"` | Container of toggle options. |


## Dependencies

### Depends on

- [kv-toggle-button](../toggle-button)

### Graph
```mermaid
graph TD;
  kv-toggle-switch --> kv-toggle-button
  kv-toggle-button --> kv-tooltip
  kv-toggle-button --> kv-radio
  kv-toggle-button --> kv-icon
  kv-tooltip --> kv-portal
  kv-tooltip --> kv-tooltip-text
  style kv-toggle-switch fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


