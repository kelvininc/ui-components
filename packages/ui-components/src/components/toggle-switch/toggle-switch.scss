@import '../../assets/styles/globals';

.toggle-switch-container {
	display: flex;

	> :last-child {
		--border-radius-top-right: 4px;
		--border-radius-bottom-right: 4px;
	}

	> :first-child {
		--border-radius-top-left: 4px;
		--border-radius-bottom-left: 4px;
	}

	kv-toggle-button {
		--border-radius-top-right: 0;
		--border-radius-bottom-right: 0;
		--border-radius-top-left: 0;
		--border-radius-bottom-left: 0;

		--button-padding-small: #{$spacing $spacing-3x};
		--button-padding-large: #{$spacing-2x $spacing-3x};

		--text-color-default: #{kv-color('neutral-6')};
		--text-color-active: #{kv-color('neutral-0')};
		--text-color-disabled: #{kv-color('neutral-4')};
		--text-color-disabled-active: #{kv-color('neutral-4')};

		--background-color-default: transparent;
		--background-color-active: #{kv-color('primary')};
		--background-color-disabled: transparent;
		--background-color-disabled-active: #{kv-color('neutral-1')};

		--border-color-default: #{kv-color('neutral-3')};
		--border-color-active: #{kv-color('primary')};
		--border-color-disabled: #{kv-color('neutral-2')};
		--border-color-disabled-active: #{kv-color('neutral-2')};

		&::part(toggle-label) {
			@include kv-font-label-small-regular();
		}
	}
}
