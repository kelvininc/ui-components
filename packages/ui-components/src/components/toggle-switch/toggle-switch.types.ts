import { EventEmitter } from '@stencil/core';
import { EComponentSize, IToggleButton } from '../../types';

export type IToggleSwitchOption<T = string | number> = Omit<IToggleButton<T>, 'withRadio'>;

export interface IToggleSwitch<T extends string | number | symbol = string | number | symbol> {
	/** (optional) List of toggle switch options */
	options: IToggleSwitchOption<T>[];
	/** (optional) If `true` all toggle buttons will be disabled */
	disabled?: boolean;
	/** (optional) Toggle switch selected option key */
	selectedOption?: T;
	/** (optional) A record with the button's key and its individual disabled state */
	disabledButtons?: Record<T, boolean>;
	/** (optional) Sets the size for all toggle buttons */
	size?: EComponentSize;
}

export interface IToggleSwitchEvents {
	/** When the toggle switch selection changes, emit the requested value */
	checkedChange: EventEmitter<string>;
}
