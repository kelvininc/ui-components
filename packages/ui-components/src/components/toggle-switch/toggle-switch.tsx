import { Component, h, Prop, Host, Event, EventEmitter } from '@stencil/core';
import { IToggleSwitch, IToggleSwitchEvents, IToggleSwitchOption } from './toggle-switch.types';
import { EComponentSize } from '../../types';

/**
 * @part toggle-switch-option-container - Container of toggle options.
 */
@Component({
	tag: 'kv-toggle-switch',
	styleUrl: 'toggle-switch.scss',
	shadow: true
})
export class KvToggleSwitch implements IToggleSwitch, IToggleSwitchEvents {
	/** @inheritdoc */
	@Prop({ reflect: true }) options: IToggleSwitchOption[] = [];
	/** @inheritdoc */
	@Prop({ reflect: true }) disabled?: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) size: EComponentSize = EComponentSize.Small;
	/** @inheritdoc */
	@Prop({ reflect: true }) selectedOption?: string = '';
	/** @inheritdoc */
	@Prop({ reflect: true }) disabledButtons?: Record<string, boolean> = {};

	/** @inheritdoc */
	@Event() checkedChange: EventEmitter<string>;

	render() {
		return (
			<Host>
				<div class="toggle-switch-container">
					{this.options.map(option => (
						<kv-toggle-button
							part="toggle-switch-option-container"
							exportparts="toggle-button"
							icon={option.icon}
							value={option.value}
							label={option.label}
							tooltip={option.tooltip}
							size={this.size ?? option.size}
							preventDefault={option.preventDefault}
							checked={this.selectedOption === option.value}
							disabled={this.disabled || this.disabledButtons[option.value] || option.disabled}
						/>
					))}
				</div>
			</Host>
		);
	}
}
