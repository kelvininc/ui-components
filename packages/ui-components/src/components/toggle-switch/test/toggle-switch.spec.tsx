import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { cloneDeep } from 'lodash-es';
import { KvToggleSwitch } from '../toggle-switch';
import { TOGGLE_BUTTON_ITEMS } from './toggle-switch.mock';
import { h } from '@stencil/core';

describe('Toggle Switch (unit tests)', () => {
	let page: SpecPage;
	let component: KvToggleSwitch;

	const toggleOptionsMock = cloneDeep(TOGGLE_BUTTON_ITEMS);

	describe('when rendering with default props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvToggleSwitch],
				template: () => <kv-toggle-switch options={toggleOptionsMock}></kv-toggle-switch>
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should set the correct component tabs', () => {
			expect(component.options).toEqual([
				{
					value: 'opt1',
					label: 'Option 1'
				},
				{
					value: 'opt2',
					label: 'Option 2'
				},
				{
					value: 'opt3',
					label: 'Option 3'
				},
				{
					value: 'opt4',
					label: 'Option 4'
				}
			]);
		});
	});
});
