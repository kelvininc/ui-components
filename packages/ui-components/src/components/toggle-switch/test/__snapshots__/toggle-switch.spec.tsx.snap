// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle Switch (unit tests) when rendering with default props should match the snapshot 1`] = `
<kv-toggle-switch selected-option="" size="small">
  <mock:shadow-root>
    <div class="toggle-switch-container">
      <kv-toggle-button exportparts="toggle-button" label="Option 1" part="toggle-switch-option-container" size="small" value="opt1"></kv-toggle-button>
      <kv-toggle-button exportparts="toggle-button" label="Option 2" part="toggle-switch-option-container" size="small" value="opt2"></kv-toggle-button>
      <kv-toggle-button exportparts="toggle-button" label="Option 3" part="toggle-switch-option-container" size="small" value="opt3"></kv-toggle-button>
      <kv-toggle-button exportparts="toggle-button" label="Option 4" part="toggle-switch-option-container" size="small" value="opt4"></kv-toggle-button>
    </div>
  </mock:shadow-root>
</kv-toggle-switch>
`;
